src/lib/y2storage/package_handler.rb

```diff
@@ -0,0 +1,168 @@
+  # storage-related packages to the set of packages to install during
+  # installation, install small sets of packages in the installed system (for
+  # the expert partitioner) as needed.
+  #
+  class PackageHandler
+    include Yast::Logger
+    include Yast::I18n
+
+    attr_reader :pkg_list
+
+    PROPOSAL_ID = "storage_proposal"
+
+    def initialize
+      textdomain("storage-ng")
+      reset
+    end
+
+    # Clear the package list
+    #
+    def reset
+      @pkg_list = []
+    end
+
+    # Add a number of packages to the list of packages to be installed
+    #
+    # @param  [Array<String>] package names
+    # @return [Array<String>] new package list (may contain duplicates)
+    #
+    def add_packages(pkg_list)
+      @pkg_list.concat(pkg_list)
+    end
+
+    # Add the packages for the storage features in 'devicegraph' to the list of
+    # packages to be installed
+    #
+    # @param [::Storage::devicegraph] devicegraph to obtain the features from
+    # @return [Array<String>] new package list (may contain duplicates)
+    #
+    def add_feature_packages(devicegraph)
+      used_features = UsedStorageFeatures.new(devicegraph)
+      add_packages(used_features.feature_packages)
+    end
+
+    # Start a package dependency resolver run
+    #
+    def solve
+      Yast::Pkg.PkgSolve(true)
+    end
+
+    # Commit the changes depending on the current mode: During OS installation,
+    # mark the package list for installation. In the installed system, install
+    # them immediately.
+    #
+    def commit
+      if Yast::Mode.installation
+        set_proposal_packages
+      else
+        install!
+      end
+    end
+
+    # Execute package installation. This will install the stored package list
+    # immediately, so it is not advisable to do this during the OS
+    # installation. In the latter case, use 'set_proposal_packages' instead.
+    #
+    # @return true on success, false on error
+    #
+    def install!
+      compact
+      return if @pkg_list.empty?
+      log.info("Installing #{pkg_list}")
+      success = Yast::Package.DoInstall(@pkg_list)
+      if !success
+        log.error("ERROR: Some packages could not be installed")
+        install_error_popup
+      end
+      success
+    end
+
+    # Set the proposal packages for storage. This marks the packages for
+    # installation; it does not install them yet.
+    #
+    def set_proposal_packages
+      compact
+      return if @pkg_list.empty?
+      log.info("Marking #{pkg_list} for installation")
+      if !Yast::PackagesProposal.SetResolvables(PROPOSAL_ID, :package, @pkg_list)
+        log.error("PackagesProposal::SetResolvables() for #{pkg_list} failed")
+        set_resolvables_error_popup
+      end
+      solve
+      nil
+    end
+
+    # Remove duplicates from the package list and those packages that are
+    # already installed.
+    #
+    # @return [Array<String>] compacted list
+    #
+    def compact
```

Comment:
I'm not against removing duplicates since that looks safe to me (although I would do it while adding the packages to the list). I'm worried about the second line in `#compact`, not about the first one.

I just feel there is something intrinsically wrong in conditioning the behavior if the object to the state of the system the first time `#compact` was called (even subsequent calls to `#compact` are affected). Specially since the method is called indirectly in several places. I think this is more correct than having a `#compact` method.
```ruby
def compact_pkg_list
  list = @pkg_list.uniq # To be honest, I would do this in #add_packages, but I don't care much
  list.reject {|pkg| Yast::Package.Installed(pkg) }
end
```

An this is YaST, we don't need "anybody from outside" to create insane workflows to shoot our feet. Just think about
```ruby
# Assume this adds package "xx"
pkg_handler.add_feature_packages(devicegraph)
if !pkg_handler.commit
   log.info("Commit of packages failed. Let's try to fix it")
   do_something
   # The following call has the side effect of uninstalling "xx". This kind of side-effects are quite common in YaST
   Yast::SomeModule.DoAnotherThing()
   do_something_else
   log.info("Retrying commit.")
   # "xx" not installed, although now it makes sense
   pkg_handler.commit
end
```

I will not be that surprised if we end up having code like that in the future. And I won't be surprised of the programmer using this API overlooks the mentioned problem (I myself would assume that every call to `#commit` is just based in the current system state).